### ISSUE TITLE
file.createReadStream & all dependencies return potentially broken data with gzip/validation stream on

### ISSUE THREAD

**Issue by AVVS** [googleapis/nodejs-storage#108]
#### Environment details

  - OS: any
  - Node.js version: any
  - npm version: 5
  - @google-cloud/storage version: >= 1.4.0

#### Steps to reproduce

https://github.com/googleapis/nodejs-storage/blob/3921796b26b15abccfba0556f05962fbb617870e/src/file.js#L584-L592

It's possible for 'end' of rawResponseStream to fire before all of the data has been processed by gzip / validation stream and end transformStream prematurely, effectively loosing data in the process.

'end' must only be called after it has been confirmed that further stream had emitted 'end' event

https://github.com/googleapis/nodejs-storage/blob/3921796b26b15abccfba0556f05962fbb617870e/src/file.js#L594-L609

so basically something like can return incomplete data:

`file.createReadStream('name-of-file.txt').pipe(transform).pipe(fs.createWriteStream())`

Essentially gunzip will take sometime, will be performed automatically, but 'end' will fire before all the data has been passed further to transform resulting in data corruption

**Comment by stephenplusplus** [comment]
Thanks for catching this. I'll get a fix going ASAP.

**Comment by stephenplusplus** [comment]
PR sent in #110.

**Comment by aiham** [comment]
@AVVS Thanks for reporting this.

@stephenplusplus Unfortunately I lost data because of this bug. What would you recommend to avoid data loss from this type of issue? Should I be doing my own custom checksum validation? I knew `File.prototype.createWriteStream` already does validation by default so it seemed redundant to do it again but it turns out it might be worth it after all.

**Comment by stephenplusplus** [comment]
@aiham as it stands, we consider this bug fixed. If you're still having the same issue after upgrading to 1.5.2, could you please open a new issue and include a reproduction case?

**Comment by aiham** [comment]
@stephenplusplus Thanks for the response.

I'm not experiencing this bug after upgrading to 1.5.2, I was asking for best practice suggestions to mitigate future issues since #110 contains no new test cases to avoid the regression of a p0 bug.
